ArduPlane/mode_loiter.cpp

```diff
@@ -40,26 +40,34 @@ void ModeLoiter::update()
 bool ModeLoiter::isHeadingLinedUp(const Location loiterCenterLoc, const Location targetLoc)
 {
     // Return true if current heading is aligned to vector to targetLoc.
-    // Tolerance is initially 10 degrees and grows at 10 degrees for each loiter circle completed.
 
-    if (loiterCenterLoc.get_distance(targetLoc) < 1.05f * fabsf(plane.loiter.radius)) {
-        /* Whenever next waypoint is within the loiter radius plus 5%,
+    const float targetDistanceFromCenter = loiterCenterLoc.get_distance(targetLoc);
+    if (targetDistanceFromCenter < 0.99f * fabsf(plane.loiter.radius)) {
+        /* Whenever next waypoint is within the loiter radius (damped slightly for rounding error),
            maintaining loiter would prevent us from ever pointing toward the next waypoint.
            Hence break out of loiter immediately
          */
         return true;
+    } else {
+        // Bearing in centi-degrees
+         const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
+        if (targetDistanceFromCenter < 1.05f * fabsf(plane.loiter.radius)) {
+            /*  Target Waypoint is between [radius, 1.05*radius) to center so we can
+                point toward the next waypoint, but there may be overshooting. Hence,
+                tolerance increased to 20 degrees on initial check to reduce loop-backs.
+            */
+            return isHeadingLinedUp_cd(bearing_cd, 2000);
+        }
+        // Target waypoint is far enough from center, base tolerance is acceptable.
+        return isHeadingLinedUp_cd(bearing_cd);
     }
-
-    // Bearing in centi-degrees
-    const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
-    return isHeadingLinedUp_cd(bearing_cd);
 }
 
 
-bool ModeLoiter::isHeadingLinedUp_cd(const int32_t bearing_cd)
+bool ModeLoiter::isHeadingLinedUp_cd(const int32_t bearing_cd, const int32_t initial_bearing_margin)
 {
     // Return true if current heading is aligned to bearing_cd.
-    // Tolerance is initially 10 degrees and grows at 10 degrees for each loiter circle completed.
+    // Tolerance is initially "initial_bearing_margin" and grows at 10 degrees for each loiter circle completed.
 
     // get current heading.
     const int32_t heading_cd = (wrap_360(degrees(ahrs.groundspeed_vector().angle())))*100;
```

Comment:
not your problem, but we don't need to do the wrap here because we wrap the final calculation on 75.
```suggestion
    const int32_t heading_cd = degrees(ahrs.groundspeed_vector().angle())*100;
```